### DISCUSSION TITLE
How can I show spinner while doing backend API call?

### DISCUSSION THREAD

**Discussion by jackyzsn** [emilk/egui#1809] [Q&A]
I have a button click it's connecting the database. I want to show a spinner say on the status bar on the bottom of the window, when the button is clicked and connecting the MongoDB. I have a ui enable component to control the spinner show or hide. Before the API call, I set the variable to true and hoping the spinner will show. After the API call, I set it back to false so the spinner will be hidden. But it's not working. When the API is trying to connect, the UI is not refreshing I guess. How can I do this?                 

               ui.add(password::password(encrypt_key));

                if ui.button("📎 Connect").clicked() {
                    *in_progress = true;

                    status.replace_range(.., &format!("Retrieving notes from {:?}..", run_mode));
                    let mut has_error = false;
                    let mut api_result = Vec::new();

                    if *run_mode == RunMode::Cloud {
                        tokio::runtime::Builder::new_multi_thread()
                            .enable_all()
                            .build()
                            .unwrap()
                            .block_on(async {
                                api_result = retrieve_all_note_summaries(
                                    note_group.to_string(),
                                    note_user_id.to_string(),
                                    encrypt_key.to_string(),
                                )
                                .await
                                .unwrap_or_else(|err| {
                                    status.replace_range(
                                        ..,
                                        &format!("{} {}", "Connection failed ..", err.to_string()),
                                    );
                                    has_error = true;

                                    Vec::new()
                                });
                            });

**Comment by Titoot** [comment]
_i know this is quite old but for anyone else searching for a solution_

I think async operation in update isn't recommended which results in the UI hanging
so that's my "bad" solution to the problem:
i created a static in the API class with a Mutex with an is_loading arbitrate 
```rust
lazy_static! {
    pub static ref UI_STATE: Mutex<UiState> = Mutex::new(UiState {
        is_loading: false,
    });
}
```
then used the ui_state in ui code
```rust
egui::CentralPanel::default().show(ctx, |ui: &mut egui::Ui| {
            let ui_state = UI_STATE.lock().unwrap();
            self.is_enabled = !ui_state.is_loading;
            if ui_state.is_loading {
                ui.horizontal(|ui| {
                    ui.label("Loading... Please wait.");
                    ui.spinner();
                });
            }
            ui.add_enabled_ui(self.is_enabled, |ui| {
                self.show_main_ui(ui);
            });
        });
```
and is_enabled used here to disable all UI elements until the request is done

**Comment by jafischer** [comment]
Another solution is to use the [poll-promise](https://crates.io/crates/poll-promise) crate. Then you can call async functions directly from your update method. Example:
```
    fn update(&mut self, egui_ctx: &Context, _: &mut Frame) {
...
            match &self.state {
                Created => {
                    self.log.push_str("Connecting...\n");
                    self.state = Connecting(Promise::spawn_async(connect(...)));
                }
                Connecting(poll_promise) => {
                    if let Some(result) = poll_promise.ready() {
                        match result {
                            Ok(xxx) => {
                                self.log.push_str("Doing something else...\n");
                                self.state = DoingSomethingElse(Promise::spawn_async(do_something_else(...)));
                            }
                            Err(e) => {
                                self.log.push_str(&format!("Error: {}\n", e));
                                self.state = Error;
                            }
                        }
                    }
                }
...
```
As for the lack of updating, when you have things going on in the background, you can force updates via `request_repaint`:
```
            if self.state.needs_repaint() {
                // Could also use request_repaint_after(duration) to save cpu...
                ui.ctx().request_repaint();
            }
```
